litgpt/data/text_files.py

```diff
@@ -0,0 +1,160 @@
+                inputs=train_files,
+                output_dir=str(self.data_path_train),
+                num_workers=num_workers,
+                chunk_bytes="50MB",
+            )
+        if not Path(self.data_path_val).is_dir():
+            optimize(
+                fn=partial(tokenize, tokenizer=self.tokenizer),
+                inputs=[val_files] if not isinstance(val_files, list) else val_files,
+                output_dir=str(self.data_path_val),
+                num_workers=1,  # there's only 1 file
+                chunk_bytes="50MB",
+            )
+
+    def train_dataloader(self) -> DataLoader:
+        from litdata.streaming import StreamingDataLoader, StreamingDataset, TokensLoader
+
+        train_dataset = StreamingDataset(
+            input_dir=str(self.data_path_train),
+            item_loader=TokensLoader(block_size=self.max_seq_length),
+            shuffle=True,
+            drop_last=True,
+        )
+        if self.num_workers is None:
+            num_workers = get_half_workers()
+
+        train_dataloader = StreamingDataLoader(
+            train_dataset, batch_size=self.batch_size, pin_memory=True, num_workers=num_workers, drop_last=True
+        )
+        return train_dataloader
+
+    def val_dataloader(self) -> DataLoader:
+        from litdata.streaming import StreamingDataset, TokensLoader
+
+        if self.num_workers is None:
+            num_workers = get_half_workers()
+
+        val_dataset = StreamingDataset(
+            input_dir=str(self.data_path_val),
+            item_loader=TokensLoader(block_size=self.max_seq_length),
+            shuffle=True,
+            # Consider setting to False, but we would lose some samples due to truncation when world size > 1
+            drop_last=True,
+        )
+        val_dataloader = DataLoader(
+            val_dataset, batch_size=self.batch_size, pin_memory=True, num_workers=num_workers, drop_last=True
+        )
+        return val_dataloader
+
+
+def tokenize(filename: str, tokenizer: Tokenizer):
+    if tokenizer is None:
+        raise ValueError(
+            "Tokenizer is None. If you are using this data module via `litgpt pretrain`, "
+            "please provide a valid `--tokenizer_dir` path."
+        )
+    with open(filename, "r", encoding="utf-8") as file:
+        text = file.read()
+    text = text.strip()
+
+    chunks = []
+    total_length = len(text)
+    num_chunks = 10
+    chunk_size = total_length // num_chunks
+    for i in range(num_chunks):
+        start_index = i * chunk_size
+        end_index = (i + 1) * chunk_size if i < 9 else total_length
+        chunks.append(text[start_index:end_index])
+
+
+    global_rank = int(os.environ["DATA_OPTIMIZER_GLOBAL_RANK"])
+    num_workers = int(os.environ["DATA_OPTIMIZER_NUM_WORKERS"])
+    local_rank = global_rank % num_workers
+    for example in tqdm(chunks, position=local_rank):
+        tokens = tokenizer.encode(example.strip(), bos=True, eos=False)  # encode the text, use BOS
```

Comment:
I don't understand some of the assumptions here. 

1. How can we insert a BOS, when we are in the middle of the document? Isn't one text file = one document? Note that in the TinyStories this is not a problem, because one entry in the json file = one story so the bos/eos can be set. But here, I'm not convinced. In my opinion, if we are doing BOS we should insert it only at the very beginning, or if using EOS add it at the very end of the document.

2. Isn't the whitespace part of the document and meaningful? I would prefer if we assumed the data was already cleaned beforehand, and not do any stripping.

Reply:
This was bad/lazy. I removed this and added a note that we expect users to bring preprocessed data. We can later maybe have tutorials or tools for preprocessing.